aggsender/aggsender.go

```diff
@@ -416,24 +428,120 @@ func (a *AggSender) sendCertificate(ctx context.Context) (*agglayertypes.Certifi
 	return certificate, nil
 }
 
-// callValidator calls the validator to validate the certificate
-func (a *AggSender) callValidator(
+// pollValidators committee signers to validate the certificate and gather signatures
+func (a *AggSender) pollValidators(
 	ctx context.Context,
+	validators []types.CertificateValidateAndSigner,
+	signaturesThreshold uint32,
 	certificate *agglayertypes.Certificate,
-	lastL2BlockInCert uint64) ([]byte, error) {
-	if a.validator == nil {
+	lastL2BlockInCert uint64,
+) ([][]byte, error) {
+	if len(validators) == 0 {
+		a.log.Warnf("skipping certificate validation, because there are no validators configured")
+
 		return nil, nil
 	}
+
 	a.log.Infof("delegating certificate validation: %s", certificate.Brief())
-	validatorSignature, err := a.validator.ValidateAndSignCertificate(ctx, certificate, lastL2BlockInCert)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	sigCh := make(chan []byte, len(validators))
+	errCh := make(chan error, len(validators))
+
+	g, gctx := errgroup.WithContext(ctx)
+
+	for _, v := range validators {
+		g.Go(func() error {
+			status, err := v.HealthCheck(ctx)
+			if err != nil {
+				a.log.Warnf("error checking validator health (URL=%s): %v", v.URL(), err)
+			}
+
+			if !status.IsHealthy() {
+				a.log.Warnf("validator status (URL=%s) is not healthy: %s", v.URL(), status.String())
+				return nil // skip the unhealthy validator
+			}
+
+			a.log.Infof("validator health check (URL=%s): %s", v.URL(), status.String())
+
+			sig, err := v.ValidateAndSignCertificate(gctx, certificate, lastL2BlockInCert)
+			if err != nil {
+				a.log.Errorf("validator %v failed to validate the certificate: %v", v, err)
+				errCh <- err
+				return nil // don't fail the group
+			}
+
+			select {
+			case sigCh <- sig:
+				if uint32(len(sigCh)) >= signaturesThreshold {
+					cancel() // early stop, because threshold is reached
+				}
+			case <-gctx.Done():
+				return nil
+			}
+			return nil
+		})
+	}
+
+	// Wait for all goroutines to finish in background
+	go func() {
+		_ = g.Wait()
+		close(sigCh)
+		close(errCh)
+	}()
+
+	signatures := make([][]byte, 0, len(validators))
+	for sig := range sigCh {
+		signatures = append(signatures, sig)
+		if uint32(len(signatures)) >= signaturesThreshold {
+			cancel()
+			break
+		}
+	}
+
+	// Collect all errors from errCh
+	errs := make([]error, 0)
+	for err := range errCh {
+		errs = append(errs, err)
+	}
+
+	// If we didn’t reach threshold, return combined error
+	if uint32(len(signatures)) < signaturesThreshold {
+		if len(errs) > 0 {
+			return signatures, errors.Join(errs...)
```

Comment:
I would perhaps log the errors returned by validators even if we have a threshold. WDYT?

Reply:
I see you are logging them in the go routine.